File: 21-CountingMinutesI/21.js
// Given a time range string, find the number of minutes between the two times

function CountingMinutesI(str) { 
  //Format is [all,hr,min,period,hr,min,period,...]
  var data = /([0-9]{1,2}):([0-9]{1,2})(am|pm)-([0-9]{1,2}):([0-9]{1,2})(am|pm)/.exec(str); 
  
  var time1 = (parseInt(data[1])*60 % 60*12) + parseInt(data[2]) + (data[3] === "pm" ? 60*12 : 0);
  var time2 = (parseInt(data[4])*60 % 60*12) + parseInt(data[5]) + (data[6] === "pm" ? 60*12 : 0);

  return (60*24 + time2 - time1) % 60*24; //1440 mins in a day
}

console.log(CountingMinutesI("11:59pm-12:00am") === 1);
console.log(CountingMinutesI("5:20pm-6:30pm") === 70);
console.log(CountingMinutesI("12:30pm-12:00am") === 690);
console.log(CountingMinutesI("1:23am-1:08am") === 1425);

File: 21-CountingMinutesI/21.hs
import Data.List.Split
import Text.Regex.Posix

countingMinutesI :: String -> Int --Without regex
countingMinutesI str = mod (60*24 + time2 - time1) 60*24--1440 mins in a day
  where splitTimes = (\(x:y:[]) -> (x,y)) . (splitOn "-")
        time1      = parseTime . fst . splitTimes $ str
        time2      = parseTime . snd . splitTimes $ str
        parseTime time = hr + min + period
            where hr     = ((read $ takeWhile (/= ':') time) `mod` 12) * 60 --Parse first 1-2 digts
                  min    = read . take 2 . drop 1 . dropWhile(/= ':') $ time --Parse next 2 after ':'
                  period = if (elem 'a' time) then 0 else 60*12 --Check for am/pm

countingMinutesIRegex :: String -> Int --With regex!
countingMinutesIRegex str = mod (60*24 + time2 - time1) 60*24 --1440 mins in a day
  where splitTimes = (\(x:y:[]) -> (x,y)) . (splitOn "-")
        time1      = parseTime . fst . splitTimes $ str
        time2      = parseTime . snd . splitTimes $ str
        parseTime time =
          let (_,_,_,[hr,min,period]) = time =~ "([0-9]{1,2}):([0-9]{1,2})(am|pm)" :: (String,String,String,[String]) --yay,regex
          in (read hr `mod` 12) * 60 + (read min) + (if (elem 'a' period) then 0 else 60*12) -- modded hrs + mins + period

main = do
  print $ countingMinutesI "11:59pm-12:00am" == 1
  print $ countingMinutesI "5:20pm-6:30pm" == 70
  print $ countingMinutesI "12:30pm-12:00am" == 690
  print $ countingMinutesI "1:23am-1:08am" == 1425
  print $ countingMinutesIRegex "11:59pm-12:00am" == 1
  print $ countingMinutesIRegex "5:20pm-6:30pm" == 70
  print $ countingMinutesIRegex "12:30pm-12:00am" == 690
  print $ countingMinutesIRegex "1:23am-1:08am" == 1425